aws_advanced_python_wrapper/limitless_connection_plugin.py

```diff
@@ -0,0 +1,543 @@
+                context.get_limitless_routers(), HostRole.WRITER, "weighted_random")
+            logger.debug("LimitlessRouterServiceImpl.SelectedHost", "None" if selected_host_info is None else selected_host_info.host)
+        except Exception as e:
+            if self._is_login_exception(e):
+                raise e
+
+            self._retry_connection_with_least_loaded_routers(context)
+            return
+
+        if selected_host_info is None:
+            self._retry_connection_with_least_loaded_routers(context)
+            return
+
+        try:
+            context.set_connection(self._plugin_service.connect(selected_host_info, context.get_props()))
+        except Exception as e:
+            if self._is_login_exception(e):
+                raise e
+
+            if selected_host_info is not None:
+                logger.debug("LimitlessRouterServiceImpl.FailedToConnectToHost", selected_host_info.host)
+                selected_host_info.set_availability(HostAvailability.UNAVAILABLE)
+
+            self._retry_connection_with_least_loaded_routers(context)
+
+    def _get_limitless_routers(self, cluster_id: str, props: Properties) -> List[HostInfo]:
+        # Convert milliseconds to nanoseconds
+        cache_expiration_nano: int = WrapperProperties.MONITOR_DISPOSAL_TIME_MS.get_int(props) * 1_000_000
+        self._limitless_router_cache.set_cleanup_interval_ns(cache_expiration_nano)
+        routers = self._limitless_router_cache.get(cluster_id)
+        if routers is None:
+            return []
+        return routers
+
+    def _retry_connection_with_least_loaded_routers(self, context: LimitlessConnectionContext) -> None:
+        retry_count = 0
+        max_retries = WrapperProperties.MAX_RETRIES.get_int(context.get_props())
+        while retry_count < max_retries:
+            retry_count += 1
+            if context.get_limitless_routers() is None or len(context.get_limitless_routers()) == 0 or not context.is_any_routers_available():
+                self._synchronously_get_limitless_routers_with_retry(context)
+
+                if (context.get_limitless_routers() is None
+                        or len(context.get_limitless_routers()) == 0
+                        or not context.is_any_routers_available()):
+                    logger.debug("LimitlessRouterServiceImpl.NoRoutersAvailableForRetry")
+
+                    if context.get_connection() is not None:
+                        return
+                    else:
+                        try:
+                            context.set_connection(context.get_connect_func()())
+                            return
+                        except Exception as e:
+                            if self._is_login_exception(e):
+                                raise e
+
+                            raise AwsWrapperError(Messages.get_formatted("LimitlessRouterService.UnableToConnectNoRoutersAvailable"))
```

Comment:
I think this message expects the host, we can also add the error
```suggestion
                            raise AwsWrapperError(Messages.get_formatted("LimitlessRouterService.UnableToConnectNoRoutersAvailable", context.get_host_info().host)) from e
```